pkg/querier/tenantfederation/merge_queryable.go

```diff
@@ -215,35 +221,54 @@ func (m *mergeQuerier) Close() error {
 	return errs.Err()
 }
 
+type selectJob struct {
+	seriesSet *storage.SeriesSet
+	querier   storage.Querier
+	tenantID  string
+}
+
 // Select returns a set of series that matches the given label matchers. If the
 // tenantLabelName is matched on it only considers those queriers matching. The
 // forwarded labelSelector is not containing those that operate on
 // tenantLabelName.
 func (m *mergeQuerier) Select(sortSeries bool, hints *storage.SelectHints, matchers ...*labels.Matcher) storage.SeriesSet {
 	matchedTenants, filteredMatchers := filterValuesByMatchers(defaultTenantLabel, m.tenantIDs, matchers...)
+	var jobs = make([]interface{}, len(matchedTenants))
 	var seriesSets = make([]storage.SeriesSet, len(matchedTenants))
-	var ssPos int
-	var wg sync.WaitGroup
+	var jobPos int
 	for tenantPos := range m.tenantIDs {
 		if _, matched := matchedTenants[m.tenantIDs[tenantPos]]; !matched {
 			continue
 		}
-		wg.Add(1)
-		go func(ssPos, tenantPos int) {
-			defer wg.Done()
-			seriesSets[ssPos] = &addLabelsSeriesSet{
-				upstream: m.queriers[tenantPos].Select(sortSeries, hints, filteredMatchers...),
-				labels: labels.Labels{
-					{
-						Name:  defaultTenantLabel,
-						Value: m.tenantIDs[tenantPos],
-					},
+		jobs[jobPos] = &selectJob{
+			seriesSet: &seriesSets[jobPos],
+			querier:   m.queriers[tenantPos],
+			tenantID:  m.tenantIDs[tenantPos],
+		}
+		jobPos++
+	}
+
+	run := func(ctx context.Context, jobIntf interface{}) error {
+		job, ok := jobIntf.(*selectJob)
+		if !ok {
+			return fmt.Errorf("unexpected type %T", jobIntf)
+		}
+		*job.seriesSet = &addLabelsSeriesSet{
+			upstream: job.querier.Select(sortSeries, hints, filteredMatchers...),
+			labels: labels.Labels{
+				{
+					Name:  defaultTenantLabel,
+					Value: job.tenantID,
 				},
-			}
-		}(ssPos, tenantPos)
-		ssPos++
+			},
+		}
+		return nil
+	}
+
+	err := concurrency.ForEach(m.ctx, jobs, maxConcurrency, run)
+	if err != nil {
+		panic(fmt.Sprintf("this should never happen: %v", err))
```

Comment:
This is something which is not very elegant, but I think I don't really have a choice here